### PR TITLE
[MU4] Fix #12078: Export Audio using Kontakt doesn't work and leads to crash

### LINKED ISSUE #12078: [MU4 Issue] [VST] Export Audio using Kontakt doesn't work and leads to crash
https://user-images.githubusercontent.com/90187801/174341923-f42fb45e-6779-4ced-a149-4d293f4fa12d.mp4

**Platform information**
macOS, Windows

### PR DESCRIPTION
**Pull request by RomanPudashkin**
Resolves: https://github.com/musescore/MuseScore/issues/12078

### CODE REVIEW

File: src/framework/vst/internal/vstaudioclient.cpp [hunk 1]
```diff
@@ -140,6 +140,15 @@ audio::samples_t VstAudioClient::process(float* output, audio::samples_t samples
 
     m_processData.numSamples = samplesPerChannel;
 
+    //! NOTE: From the VST3 documentation:
+    //!
+    //! Note that the ProcessData->numSamples
+    //! which indicates how many samples are used in a process call can change from call to call,
+    //! but never bigger than the maxSamplesPerBlock
+    if (samplesPerChannel != m_samplesInfo.maxSamplesPerBlock) {
```
Comment (igorkorsukov) [inline src/framework/vst/internal/vstaudioclient.cpp:148]: why not `samplesPerChannel < m_samplesInfo.maxSamplesPerBlock`?

File: src/framework/vst/internal/vstaudioclient.cpp [hunk 2]
```diff
@@ -168,17 +177,22 @@ void VstAudioClient::flush()
 
 void VstAudioClient::setBlockSize(unsigned int samples)
 {
-    if (m_samplesInfo.samplesPerBlock == samples) {
+    if (m_samplesInfo.maxSamplesPerBlock == samples) {
```
[the remark this answers is not in the document]
Comment (RomanPudashkin) [reply src/framework/vst/internal/vstaudioclient.cpp:180]: Don't quite understand the question. This is the standard optimization for setters

File: src/framework/vst/internal/vstaudioclient.cpp (hunk 1 above)
Comment (RomanPudashkin) [reply src/framework/vst/internal/vstaudioclient.cpp:148]: maxSamplesPerBlock may be bigger or smaller depending on the current situation:

* when entering notes it is 1024
* when exporting an mp3 file it is 2048

This check optimizes such cases and not only increases maxSamplesPerBlock, but also decreases it (to store smaller buffers) depending on the situation

File: src/framework/vst/internal/vstaudioclient.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (igorkorsukov) [reply src/framework/vst/internal/vstaudioclient.cpp:180]: This is the standard optimization for setters :)
```
    if (m_samplesInfo.sampleRate == sampleRate) {
        return;
    }
```

File: src/framework/vst/internal/vstaudioclient.cpp (hunk 1 above)
Comment (igorkorsukov) [reply src/framework/vst/internal/vstaudioclient.cpp:148]: The name is confusing. `max` - usually means the maximum limit. Perhaps it means that `maxSamplesPerBlock` should be named like `samplesPerBlock`

File: src/framework/vst/internal/vstaudioclient.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (igorkorsukov) [reply src/framework/vst/internal/vstaudioclient.cpp:180]: see https://github.com/musescore/MuseScore/pull/12211/files#r910149831 same confusing

**Comment by DmitryArefiev** [comment]
Tested #12078 on Win10, Mac12, LinuxMint 20.2 - **FIXED**    

Other issues related to audio export I will separately

**Review by vpereverzev — APPROVED** [review]

**Pull request merged**